File: Parallel-Csharp/src/Parallel/Linq101/JoinOperations.cs
using System;
using System.Collections.Generic;
using System.Linq;
using Parallel.Linq101.DataSources;

namespace Parallel.Linq101;

public class JoinOperations
{
    public List<Product> GetProductList() => Products.ProductList;
    public List<Customer> GetCustomerList() => Customers.CustomerList;

    public int CrossJoinQuery()
    {
        string[] categories =
        {
            "Beverages",
            "Condiments",
            "Vegetables",
            "Dairy Products",
            "Seafood"
        };

        List<Product> products = GetProductList();

        var q = categories.Join(products, c => c, p => p.Category,
            (c, p) => Tuple.Create(c, p.ProductName)).ToList();

        foreach (var v in q)
            Console.WriteLine(v.Item1 + ": " + v.Item2);
        return 0;
    }

    public int GroupJoinQuery()
    {
        string[] categories =
        {
            "Beverages",
            "Condiments",
            "Vegetables",
            "Dairy Products",
            "Seafood"
        };

        List<Product> products = GetProductList();
        
        var q = categories.GroupJoin(products, c => c, p => p.Category,
            (c, p) => Tuple.Create(c, p.ToList()))
            .ToDictionary(t=>t.Item1,t=>t.Item2);

        foreach (var v in q)
        {
            Console.WriteLine(v.Key + ":");
            foreach (var p in v.Value)
                Console.WriteLine("   " + p.ProductName);
        }
        return 0;
    }

    public int CrossGroupJoin()
    {

        string[] categories =
        {
            "Beverages",
            "Condiments",
            "Vegetables",
            "Dairy Products",
            "Seafood"
        };

        List<Product> products = GetProductList();
        
        var q = categories.GroupJoin(products, c => c, p => p.Category,
            (c, p) => Tuple.Create(c, p)).SelectMany(t=>t.Item2,
            (t,p)=>Tuple.Create(t.Item1,p.ProductName)).ToList();
        
        foreach (var v in q)
            Console.WriteLine(v.Item2 + ": " + v.Item1);

        return 0;
    }

    public int LeftOuterJoin()
    {
        string[] categories =
        {
            "Beverages",
            "Condiments",
            "Vegetables",
            "Dairy Products",
            "Seafood"
        };

        List<Product> products = GetProductList();
        
        var q = categories.GroupJoin(products, c => c, p => p.Category,
            (c, p) => Tuple.Create(c, p)).SelectMany(t=>t.Item2.DefaultIfEmpty(null),
            (t,p)=>Tuple.Create(t.Item1,p == null ? "(No products)" : p.ProductName)).ToList();

        foreach (var v in q)
            Console.WriteLine($"{v.Item2}: {v.Item1}");

        return 0;
    }
}

File: Parallel-Java/src/parallel/linq101/JoinOperations.java
package parallel.linq101;

import org.javatuples.Pair;
import parallel.linq101.datasource.Customer;
import parallel.linq101.datasource.Data;
import parallel.linq101.datasource.Product;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

public class JoinOperations {
    public List<Product> getProductList() {
        return Data.getProductList();
    }

    public List<Customer> getCustomerList() {
        return Data.getCustomerList();
    }

    public int crossJoinQuery() {
        String[] categories = {"Beverages", "Condiments", "Vegetables", "Dairy Products", "Seafood"};
        List<Product> products = getProductList();
        var q = Arrays.stream(categories).flatMap(
                c -> products.stream().filter(p -> p.getCategory().equals(c)).map(
                        p -> Pair.with(c, p.productName)
                )).collect(Collectors.toList());
        for (var v : q)
            System.out.println(v.getValue0() + ": " + v.getValue1());
        return 0;
    }

    public int groupJoinQuery() {
        String[] categories = {"Beverages", "Condiments", "Vegetables", "Dairy Products", "Seafood"};
        List<Product> products = getProductList();
        var q = Arrays.stream(categories).map(
                c -> Pair.with(c, products.stream().filter(p -> p.getCategory().equals(c))
                        .collect(Collectors.toList()))
        ).collect(Collectors.toMap(t -> t.getValue0(), t -> t.getValue1()));
        for (var v : q.entrySet()) {
            System.out.println(v.getKey() + ":");
            for (var p : v.getValue())
                System.out.println("   " + p.getProductName());
        }
        return 0;
    }

    public int crossGroupJoin() {
        String[] categories = {"Beverages", "Condiments", "Vegetables", "Dairy Products", "Seafood"};
        List<Product> products = getProductList();
        var q = Arrays.stream(categories).map(
                c -> Pair.with(c, products.stream().filter(p -> p.getCategory().equals(c))
                        .collect(Collectors.toList()))).flatMap(t -> t.getValue1().stream()
                .map(p -> Pair.with(t.getValue0(), p.getProductName()))).collect(Collectors.toList());
        for (var v : q)
            System.out.println(v.getValue1() + ": " + v.getValue0());
        return 0;
    }

    public int leftOuterJoin() {
        String[] categories = {"Beverages", "Condiments", "Vegetables", "Dairy Products", "Seafood"};
        List<Product> products = getProductList();
        var q = Arrays.stream(categories).map(
                c -> Pair.with(c, products.stream().filter(p -> p.getCategory().equals(c))
                        .collect(Collectors.toList()))).flatMap(t -> defaultIfEmpty(t.getValue1(), null)
                .stream().map(p -> Pair.with(t.getValue0(),
                        p == null ? "(No products)" : p.getProductName()))).collect(Collectors.toList());
        for (var v : q)
            System.out.println(v.getValue1() + ": " + v.getValue0());
        return 0;
    }

    public static <T> List<T> defaultIfEmpty(List<T> list, T def) {
        if (list.isEmpty()) {
            List<T> ll = new ArrayList<T>();
            ll.add(def);
            return ll;
        } else return list;
    }

}
